### PR TITLE
Add workaround for numpy issue 12142

### LINKED ISSUE #1694: Comparison operator fails with numpy scalar
```
>>> cupy.zeros(1) > cupy.float64(1)
array([False])
>>> cupy.float64(1) < cupy.zeros(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cupy/core/core.pyx", line 1728, in cupy.core.core.ndarray.__array_ufunc__
  File "cupy/core/_kernel.pyx", line 780, in cupy.core._kernel.ufunc.__call__
  File "cupy/core/_kernel.pyx", line 81, in cupy.core._kernel._preprocess_args
TypeError: Unsupported type <class 'numpy.ndarray'>
```

It seems `__array_ufunc__` is called with 0-dim `numpy.ndarray` converted from numpy's scalar (`numpy.float32`, ...).

### PR DESCRIPTION
**Pull request by toslunar**
Close #1694.

### CODE REVIEW

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
```
>>> import numpy as np
>>> class C(object):
...  __array_priority__ = 1
...  def __gt__(self, other): print(type(other))
...
>>> np.int8(3) < C()
<class 'numpy.ndarray'>
>>> np.__version__
'1.12.1'
```
```
>>> import numpy as np
>>> class C(object):
...  __array_priority__ = 1
...  def __gt__(self, other): print(type(other))
...
>>> np.int8(3) < C()
<class 'numpy.int8'>
>>> np.__version__
'1.13.3'
```

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1720,10 +1720,21 @@ cdef class ndarray:
...
+                cp_ufunc = getattr(cupy, name)
             except AttributeError:
                 return NotImplemented
+            if name in [
+                    'greater', 'greater_equal', 'less', 'less_equal',
+                    'equal', 'not_equal']:
+                # workaround for numpy/numpy#12142
+                inputs = tuple([
+                    x.item()
+                    if isinstance(x, numpy.ndarray) and x.shape == ()
```
Comment (okuta) [inline cupy/core/core.pyx]: How about `x.ndim == 0` ?

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**